### PR TITLE
Replace instances of write[ln](format(...)) with writef(...)

### PR DESCRIPTION
**Pull request by bradcray**
Many of our tests used the format() capability prior to the existence of writef().
Now that we have writef() (and are considering deprecating format in favor
of an swritef-/string.writef-style capability), it seems good to rewrite these
tests to use writef() instead.

Here I've tackled those that I own or have perhaps inherited from other
old-time team members.  I haven't tackled those owned by current
team members (primarily Vass and Michael), and have also left a few
cases that use format() in more of a "give me a string" style.

In working on this, while I benefitted greatly from the support for ##.####
style formatting, I also suffered from it a few times, as when wanting to
do:

    writef("#%i\n", i);

which generates a runtime error because the i is matched to the # rather
than the %i as intended.  The right way to get this is:

    writef("%#%i\n", i);

I'm worried this will be confusing for C programmesr (lack of a % looks
for a match; presence of a % doesn't), so have started an email
discussion with Michael about whether we might prefer to require
%{##.###} for such cases (which I think is already supported optionally)
and deprecate the %# support.

Here are a few noteworthy test cases to call out:

* Random/deitz/test1D2D.chpl : This test was using an interesting case of
  promotion to get the array's print formatting (spaces between elements
  and linefeeds in appropriate places).  I couldn't see a way to promote
  writef(), nor am I sure that makes sense, so instead I put the loop outside
  the writef(), inserted a space before each element for simplciity and updated
  the .good files.

* Removed the testformat* tests which existed to test format().

* Ran dos2unix on intsort.mtml.chpl, making the diff bad (sorry).  I did this
  as two commits, do jumping into one of the commits should give a better
  diff (though github does a good job of highlighting the intra-line diffs as
  well.

* In the amr tests, variables had been used for the format strings, and it
  seemed that without a string literal juxtaposition capability in the language
  (which arguably we'd like but haven't implemented), this seemed hard to
  preserve, so I inlined the control strings instead, which is a little regrettable...

* The parRandom2-domBlock* futures had gotten out of date with reality
  so in addition to converting their format()s, I brought them up-to-date and
  gave them .bad files to prevent future quiet failures.

### CODE REVIEW

**Comment by bradcray** [comment]
This passed testing and is ready for review.  @Kyle-B and/or @mppf, would you like to take a look?

**Comment by mppf** [comment]
I can review this.

**Comment by mppf** [comment]
Brad, you said "In the amr tests, variables had been used for the format strings, and it
seemed that without a string literal juxtaposition capability in the language
(which arguably we'd like but haven't implemented), this seemed hard to
preserve, so I inlined the control strings instead, which is a little regrettable..."

Are the writef calls only working with string literals for the format? You should
be able to call them with two string literals that were added myformat + "      " e.g.

File: test/modules/standard/Random/deitz/test1D2D.chpl [hunk 1]
```diff
@@ -113,8 +127,10 @@ config const n: int = 8;
   var A: [{1..n} dmapped Block(rank=1,boundingBox={1..n/2})] real;
   forall (r,a) in zip(rs.iterate(A.domain), A) do
     a = r;
-  writeln(for e in A do format("#######", e.locale.id));
-  writeln(for e in A do format("#.#####", e));
+  for e in A do writef(" #######", e.locale.id);
+  writeln();
+  for e in A do writef(" #.#####", e);
+  writeln();
```
Comment (mppf) [inline test/modules/standard/Random/deitz/test1D2D.chpl]: These tests are arguably showing the case when having format() is useful even if we have string.writef().

File: test/release/examples/benchmarks/ssca2/SSCA2_Modules/analyze_torus_graphs_stencil_rep_v2.chpl [hunk 2]
```diff
@@ -69,10 +69,11 @@ module analyze_torus_graphs {
...
     writeln ( "            bound  bound" );
 
-    for d in 1 .. G.dimensions do 
-      writeln ( format ("########", d), 
-		format ("########", vertex_domain.dim(d).low),  
-		format ("########", vertex_domain.dim(d).high) );
+    for d in 1 .. G.dimensions {
+      writef ("########", d);
+      writef ("########", vertex_domain.dim(d).low);  
+      writef ("########\n", vertex_domain.dim(d).high);
```
Comment (mppf) [inline test/release/examples/benchmarks/ssca2/SSCA2_Modules/analyze_torus_graphs_stencil_rep_v2.chpl]: You should be able to make this one writef call using "%{########}%{########}%{########}\n" which would arguably be a good use of that syntax...

File: test/release/examples/benchmarks/ssca2/SSCA2_Modules/analyze_torus_graphs_stencil_rep_v3.chpl [hunk 3]
```diff
@@ -117,10 +117,11 @@ module analyze_torus_graphs {
...
     writeln ( "            bound  bound" );
 
-    for d in 1 .. G.dimensions do 
-      writeln ( format ("########", d), 
-        format ("########", vertex_domain.dim(d).low),  
-        format ("########", vertex_domain.dim(d).high) );
+    for d in 1 .. G.dimensions {
+      writef ("########", d);
+      writef ("########", vertex_domain.dim(d).low);
+      writef ("########\n", vertex_domain.dim(d).high);
```
Comment (mppf) [inline test/release/examples/benchmarks/ssca2/SSCA2_Modules/analyze_torus_graphs_stencil_rep_v3.chpl]: If you change the other, change this one too...

File: test/studies/hpcc/common/bradc/unit/parRandom2-domblock.chpl [hunk 4]
```diff
@@ -13,13 +12,13 @@ var B: [D] real;
...
 
-forall (i,r) in (A.domain, randStr1) do
+forall (i,r) in zip(A.domain, randStr1) do
   A(i) = r;
 
 for b in B do
   b = randStr2.getNext();
 
-for (i,a,b) in (D,A,B) {
+for (i,a,b) in zip(D,A,B) {
```
Comment (mppf) [inline test/studies/hpcc/common/bradc/unit/parRandom2-domblock.chpl]: This test doesn't have any format() calls. It looks like a trivial change anyway though, so it'd probably be more trouble than it's worth to pull it out of this PR.

File: test/studies/hpcc/common/bradc/unit/parRandom2-domblock64.chpl [hunk 5]
```diff
@@ -13,18 +12,18 @@ var B: [ProblemSpace] real;
 var randStr1 = new RandomStream(314159265);
 var randStr2 = new RandomStream(314159265);
 
-forall (i,r) in (A.domain, randStr1) {
-  writeln("Doing iteration ", format("#####", i), " on locale ", here.id);
+forall (i,r) in zip(A.domain, randStr1) {
+  writeln("Doing iteration %##### on locale %i\n", i, here.id);
```
Comment (mppf) [inline test/studies/hpcc/common/bradc/unit/parRandom2-domblock64.chpl]: yeah, %##### does look confusing. But the original format wouldn't have emitted a # sign, so I don't understand why the % is there? Doesn't this version generate an extra # (from %#) before printing the number?

File: test/studies/hpcc/common/bradc/unit/parRandom2-domblock64.chpl [hunk 6]
```diff
@@ -13,18 +12,18 @@ var B: [ProblemSpace] real;
...
 }
 
 for b in B do
   b = randStr2.getNext();
 
-for (i,a,b) in (ProblemSpace,A,B) {
+for (i,a,b) in zip(ProblemSpace,A,B) {
   if (a != b) then
-    writeln("mismatch at #", format("#####", i), ": ", a, " != ", b);
+    writef("mismatch at %######: %r != %r", i, a, b);
```
Comment (mppf) [inline test/studies/hpcc/common/bradc/unit/parRandom2-domblock64.chpl]: same here...

Comment (mppf) [inline test/studies/hpcc/common/bradc/unit/parRandom2-domblock64.chpl]: no I take that back... this one prints the # like it should.

File: test/studies/lsms/shemmy/s-lsms.chpl [hunk 7]
```diff
@@ -64,7 +64,7 @@ proc main() {
 
 	for itr in 0..#nIterations {
 		if itr % reportFrequency == 0 && reportFrequency <= nIterations {
-			writeln("step ", format("####", itr), "...");
+			writef("step  ###...\n", itr);
```
Comment (mppf) [inline test/studies/lsms/shemmy/s-lsms.chpl]: It seems we lost a # here - I would have expected the writef to have #### like the original

File: test/studies/lulesh/bradc/xyztuple/lulesh-dense-3tuple.chpl [hunk 8]
```diff
@@ -272,11 +272,13 @@ proc main() {
...
+    var fmtstrnum = if debug then "%1.9er" else "%1.4er";
+    var fmtstr = fmtstrnum + " " + fmtstrnum + " " + fmtstrnum + "\n";
     for i in Nodes {
-      writer.writeln(format(fmtstr, xyz[i][X]), " ", 
-                     format(fmtstr, xyz[i][Y]), " ", 
-                     format(fmtstr, xyz[i][Z]));
+      //
+      // TODO: Why doesn't this one work?
+      //
+      writer.writef(fmtstr, xyz[i][X], xyz[i][Y], xyz[i][Z]);
```
Comment (mppf) [inline test/studies/lulesh/bradc/xyztuple/lulesh-dense-3tuple.chpl]: Does this one work? It sounded from your PR message that all tests were passing. If so, maybe remove the TODO?

File: test/studies/lulesh/bradc/xyztuple/lulesh-dense-3tuple.chpl [hunk 9]
```diff
@@ -1527,20 +1529,14 @@ iter elemToNodesTuple(e) {
...
 }
 
 proc deprint(title:string, xyz:[?D] 3*real) {
   writeln(title);
   for i in D {
-    writeln(format("%3d", i), ": ", 
-            format("%3.4e", xyz[i][X]), " ", 
-            format("%3.4e", xyz[i][Y]), " ", 
-            format("%3.4e", xyz[i][Z]));
+    writef("%3n %3.4er %3.4er %3.4er\n", i, xyz[i][1], xyz[i][2], xyz[i][3]);
```
Comment (mppf) [inline test/studies/lulesh/bradc/xyztuple/lulesh-dense-3tuple.chpl]: Do you know X == 1 Y ==2 Z == 3? Not obvious to me looking at the diff....

File: test/studies/ssca2/jglewis/ssca2_version2/analyze_torus_graphs_stencil_rep_v2.chpl [hunk 10]
```diff
@@ -69,11 +69,11 @@ module analyze_torus_graphs {
...
-    for d in 1 .. G.dimensions do 
-      writeln ( format ("########", d), 
-		format ("########", vertex_domain.dim(d).low),  
-		format ("########", vertex_domain.dim(d).high) );
-
+    for d in 1 .. G.dimensions {
+      writef ("########", d);
+      writef ("########", vertex_domain.dim(d).low);
+      writef ("########\n", vertex_domain.dim(d).high);
+    }
```
Comment (mppf) [inline test/studies/ssca2/jglewis/ssca2_version2/analyze_torus_graphs_stencil_rep_v2.chpl]: This is a candidate for %{#######}

File: test/studies/ssca2/jglewis/ssca2_version2/analyze_torus_graphs_stencil_rep_v3.chpl [hunk 11]
```diff
@@ -117,11 +117,11 @@ module analyze_torus_graphs {
...
-    for d in 1 .. G.dimensions do 
-      writeln ( format ("########", d), 
-		format ("########", vertex_domain.dim(d).low),  
-		format ("########", vertex_domain.dim(d).high) );
-
+    for d in 1 .. G.dimensions {
+      writef ("########", d);
+      writef ("########", vertex_domain.dim(d).low);
+      writef ("########\n", vertex_domain.dim(d).high);
+    }
```
Comment (mppf) [inline test/studies/ssca2/jglewis/ssca2_version2/analyze_torus_graphs_stencil_rep_v3.chpl]: as is this

File: test/studies/ssca2/kristyn/analyze_torus_graphs_stencil_rep_v2.chpl [hunk 12]
```diff
@@ -69,11 +69,11 @@ module analyze_torus_graphs {
...
-    for d in 1 .. G.dimensions do 
-      writeln ( format ("########", d), 
-		format ("########", vertex_domain.dim(d).low),  
-		format ("########", vertex_domain.dim(d).high) );
-
+    for d in 1 .. G.dimensions {
+      writef ("########", d);
+      writef ("########", vertex_domain.dim(d).low);
+      writef ("########\n", vertex_domain.dim(d).high);
+    }
```
Comment (mppf) [inline test/studies/ssca2/kristyn/analyze_torus_graphs_stencil_rep_v2.chpl]: and this

File: test/studies/ssca2/kristyn/analyze_torus_graphs_stencil_rep_v3.chpl [hunk 13]
```diff
@@ -117,11 +117,11 @@ module analyze_torus_graphs {
...
-    for d in 1 .. G.dimensions do 
-      writeln ( format ("########", d), 
-		format ("########", vertex_domain.dim(d).low),  
-		format ("########", vertex_domain.dim(d).high) );
-
+    for d in 1 .. G.dimensions {
+      writef ("########", d);
+      writef ("########", vertex_domain.dim(d).low);
+      writef ("########\n", vertex_domain.dim(d).high);
+    }
```
Comment (mppf) [inline test/studies/ssca2/kristyn/analyze_torus_graphs_stencil_rep_v3.chpl]: and this

**Comment by mppf** [comment]
Looks good! You can take or leave my comments except for the one about removing a TODO comment - you really should take out the TODO if it's no longer needed.

File: test/release/examples/benchmarks/ssca2/SSCA2_Modules/analyze_torus_graphs_stencil_rep_v2.chpl (hunk 2 above)
Comment (bradcray) [inline test/release/examples/benchmarks/ssca2/SSCA2_Modules/analyze_torus_graphs_stencil_rep_v2.chpl]: Good catch!  (And, one less conversion required if we go with the %{} required approach...)

File: test/studies/hpcc/common/bradc/unit/parRandom2-domblock.chpl (hunk 4 above)
Comment (bradcray) [inline test/studies/hpcc/common/bradc/unit/parRandom2-domblock.chpl]: I call it out in the commit message -- the 64-bit version did have format calls, and required similar changes, so they seemed more akin than not.

File: test/studies/lsms/shemmy/s-lsms.chpl (hunk 7 above)
Comment (bradcray) [inline test/studies/lsms/shemmy/s-lsms.chpl]: Good catch...

File: test/studies/lulesh/bradc/xyztuple/lulesh-dense-3tuple.chpl (hunk 8 above)
Comment (bradcray) [inline test/studies/lulesh/bradc/xyztuple/lulesh-dense-3tuple.chpl]: I think this dated back to when I was struggling with C vs. Chapel format strings on that plane flight a month or so back before you set me straight.

File: test/studies/lulesh/bradc/xyztuple/lulesh-dense-3tuple.chpl (hunk 9 above)
Comment (bradcray) [inline test/studies/lulesh/bradc/xyztuple/lulesh-dense-3tuple.chpl]: I shouldn't have made that change in any case...  working too late... :P

**Comment by bradcray** [comment]
You asked "Are the writef calls only working with string literals for the format? You should
be able to call them with two string literals that were added myformat + " " e.g."

They weren't string literals, but is your point that they need not be literals (given that the compiler doesn't do anything with them anyway?).  I.e., I could do:

    var ifmt = "%5i",  rfmt="%12.6r";

and then do:

    writef(ifmt + " " + rfmt, myint, myreal);

I can imagine this would work... I'm not sure that it's preferable stylistically.  And someday we may want to warn against non-literal control strings as C does (once we have some compile-time sanity checking on them).

File: test/modules/standard/Random/deitz/test1D2D.chpl (hunk 1 above)
Comment (bradcray) [inline test/modules/standard/Random/deitz/test1D2D.chpl]: Now that the other thread has convinced me that format() ("create a string from nothing") has value beyond swritef()/string.writef() (if we can even support that), do you think I should revert this test?  What about the testformat*.chpl tests?

Comment (mppf) [inline test/modules/standard/Random/deitz/test1D2D.chpl]: I'm OK either way on these but I have a slight preference to revert the tests that used format in this way and
also leave the testformat tests there. I'll modify the testformat tests to work with whatever happens with format().

**Comment by bradcray** [comment]
@mppf: I think I finally got around to getting this into the state we want it to be in.  Do you want to give it one more review, or are you sick of it.  :)

**Comment by mppf** [comment]
Hi Brad - I looked at your commit messages - and didn't want to dig into the code again - but I'm happy with this going in any time. I'll have to adjust many of these tests in the ###.## -> %{###.##} patch though - which you could do now if you want, or leave for me.

**Pull request merged**